designs/images/cluster-controller-container-components.puml

```diff
@@ -0,0 +1,58 @@
+@startuml cluster-controller-container-components
+!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Component.puml
+
+title Component diagram for EKS-A Cluster controller
+
+AddRelTag("indirect relantionship", $textColor="orange", $lineColor="orange", $lineStyle = DashedLine())
+
+Person(user, "EKS-A user", "Manages EKS-A clusters")
+Container(kubeApiServer, "Kubernetes API server", "Go", "Provides read/write access to the kubernetes API objects stored in the etcd server and allows to subscribe to events when they change")
+Container_Ext(workloadKubeApiServer, "Workload Cluster Kubernetes API server", "Go", "Provides read/write access to the kubernetes API objects stored in the etcd server")
+Container_Ext(ociRegistry, "OCI registry", "Public ECR or other", "OCI compliant registry to store container images and Helm charts")
+
+Container_Boundary(clusteController, "Cluster Controller") {
+    AddElementTag("port", $fontColor="#ffffbf", $borderColor="#ffffbf")
+    AddElementTag("busines logic", $fontColor="#01ffff", $borderColor="#01ffff")
+
+    Component(clusterReconciler, "Cluster Reconciler", "K8s controller-runtime Reconciler", "Validates input and orchestrates Cluster reconcile requests", $tags="busines logic")
+    Component(vsphereHandler, "vSphere Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a vSphere Cluster to the desired state", $tags="busines logic")
+    Component(snowHandler, "Snow Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a Snow Cluster to the desired state", $tags="busines logic")
+
+    Component(cilimHandler, "Cilium Reconciler", "", "Reonciles the Cilium CNI installation to the desired state", $tags="busines logic")
+
+    Component(client, "Kubernetes API client", "controller-runtime client.Client", "Kubernetes client provided by controller-runtime", $tags="port")
+    Component(remoteClient, "Remote Kubernetes API client", "controller-runtime client.Client", "Kubernetes client for external cluster authenticated with kubeconfig", $tags="port")
+
+    Component(helm, "Helm client", "Helm cli binary [Go]", "Client to interact with Helm charts", $tags="port")
+
+    Rel(vsphereHandler, cilimHandler, "Delegates")
```

Comment:
Why does the individual providers delegate to cilium and not the cluster itself? Can't cluster wait for the provider to create the clusters and then delegate to the cilium controller to reconcile?

Reply:
Because the provider specific reconcilers have control over the flow of operations

We learned that providers need to hook into cluster workflows at different moments. When having a generic flow that calls into providers, you can end up with providers implementing operations they don't need. This just inverts the responsibility and gives each provider control over the high level orchestration.

Reply:
With the way things work right now, yes. But we made the same assumption on the cli and i ended up not being accurate. This just tries to avoid that.

Plus it's always easier to start with some minimal "duplication" (even though in this case I wouldn't even call it that) and then find a better common abstraction with time that try to abstract it right way and then have to split it later on.